kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h

```diff
@@ -1,1105 +1,1575 @@
+         class TDenseSpace, //= DenseSpace<double>,
+         class TLinearSolver //= LinearSolver<TSparseSpace,TDenseSpace>
+         >
+class ResidualBasedBlockBuilderAndSolverWithConstraints
+    : public BuilderAndSolver< TSparseSpace, TDenseSpace, TLinearSolver >
+{
+public:
+    ///@name Type Definitions
+    ///@{
+    KRATOS_CLASS_POINTER_DEFINITION(ResidualBasedBlockBuilderAndSolverWithConstraints);
+
+
+    typedef BuilderAndSolver<TSparseSpace, TDenseSpace, TLinearSolver> BaseType;
+
+    typedef typename BaseType::TSchemeType TSchemeType;
+
+    typedef typename BaseType::TDataType TDataType;
+
+    typedef typename BaseType::DofsArrayType DofsArrayType;
+
+    typedef typename BaseType::TSystemMatrixType TSystemMatrixType;
+
+    typedef typename BaseType::TSystemVectorType TSystemVectorType;
+
+    typedef typename BaseType::LocalSystemVectorType LocalSystemVectorType;
+
+    typedef typename BaseType::LocalSystemMatrixType LocalSystemMatrixType;
+
+    typedef typename BaseType::TSystemMatrixPointerType TSystemMatrixPointerType;
+    typedef typename BaseType::TSystemVectorPointerType TSystemVectorPointerType;
+
+    typedef Node<3> NodeType;
+
+    typedef typename BaseType::NodesArrayType NodesArrayType;
+    typedef typename BaseType::ElementsArrayType ElementsArrayType;
+    typedef typename BaseType::ConditionsArrayType ConditionsArrayType;
+
+    typedef typename BaseType::ElementsContainerType ElementsContainerType;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /**
+     * @brief Default constructor. (with parameters)
+     */
+    explicit ResidualBasedBlockBuilderAndSolverWithConstraints(
+        typename TLinearSolver::Pointer pNewLinearSystemSolver,
+        Parameters ThisParameters
+        ) : BaseType(pNewLinearSystemSolver)
+    {
+        // Validate default parameters
+        Parameters default_parameters = Parameters(R"(
+        {
+        })" );
+
+        ThisParameters.ValidateAndAssignDefaults(default_parameters);
+    }
+
+    /**
+     * @brief Default constructor.
+     */
+    explicit ResidualBasedBlockBuilderAndSolverWithConstraints(
+        typename TLinearSolver::Pointer pNewLinearSystemSolver)
+        : BaseType(pNewLinearSystemSolver)
+    {
+    }
+
+    /** Destructor.
+     */
+    ~ResidualBasedBlockBuilderAndSolverWithConstraints() override
+    {
+    }
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief Function to perform the build of the RHS. The vector could be sized as the total number
+     * of dofs or as the number of unrestrained ones
+     * @param pScheme The integration scheme considered
+     * @param rModelPart The model part of the problem to solve
+     * @param A The LHS matrix
+     * @param b The RHS vector
+     */
+    void Build(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart& rModelPart,
+        TSystemMatrixType& A,
+        TSystemVectorType& b) override
+    {

```

Comment:
@RiccardoRossi the build with constraints should be done only in case of constraints, otherwise call the base class Build, otherwise the code is being duplicated

Reply:
@loumalouomega 

could u tell me which test is failing? 'i'll check it out.

regarding your other comments, my big concern is that we are going into a major release with two versions of the builder and solver which are actually not working, at least in the case of a constraint which includes a constant.
Apparently we have a fix for the elimination builder and solver, however as i told you this gave me an unexpected bad alloc. when i have time i will try that again.
What is worst however is that the block builder and solver is not working properly, and that we are using as default in many of our testcases.

You are completely right that as of now the version of the B&S "with constraints" has duplicated code. Actually i directly copied the base class and i modified it, taking the occasion of cleaning it up.
My suggestion would be to **replace the base class** by this one, so to have one single block_builder and solver which CAN handle constraints instead of two different ones. 
This would allow to replace the base class optionally use the version of @adityaghantasala so that the work done is not lost.

On the other hand this is not really a "behaviour change" in the sense that the behaviour is identical from the user point of view. Even the user API is completely unmodified.

I do not see any of this changes been in conflict with your modification of the elimination builder and solver, that file is not included here. What is it in conflict?

Reply:
just to add that i am very convinced that without the modification of the predict, no version can work correctly.

please note that for a nonlinear problem if we have a relation of the type

     us = T * um + b

it is different to do

    Residual(um,us) = Residual(um,  us = T*um + b)

and

     Residual(um, us ) = Residual(um, us_old) - LHS*( 0  ,   T*dus)